### ISSUE TITLE
Invalid contract hinting in listen method for Illuminate/Contracts/Events/Dispatcher

### ISSUE THREAD

**Issue by EdgarSedov** [laravel/framework#35448]
- Laravel Version: 8.x

### Description:
Invalid contract for Dispatcher listen method:
https://laravel.com/api/8.x/Illuminate/Contracts/Events/Dispatcher.html
phpdoc declares what only Closure|string|null supported for $listener. 
But in fact, real implementation supports array:
https://github.com/laravel/framework/blob/8.x/src/Illuminate/Events/Dispatcher.php#L423

### Steps To Reproduce:
Make similar subscriber:
```
class CustomSubscriber
{
    public function subscribe(Dispatcher $events): void
    {
        $events->listen(CustomEvent::class, [self::class, 'handleCustomEvent']);
    }

    public function handleCustomEvent(CustomEvent $event): void
    {
        // logic
    }
}
```
When register it in events service provider and everything will work. Except u will get errors like this, if u use phpstan:
```
------ ---------------------------------------------------------- 
  Line   app/Listeners/CustomSubscriber.php             
 ------ ---------------------------------------------------------- 
  15     Parameter #2 $listener of method                          
         Illuminate\Contracts\Events\Dispatcher::listen() expects  
         Closure|string|null,                                      
         array('App\\Listeners\\CustomSubscriber',     
         'handleCustomEvent') given.                                 
 ------ ----------------------------------------------------------
```

Imho, array is much cleaner and even supported by some IDE's (renaming function name will affect string literal in listen call)

**Comment by rodrigopedra** [comment]
I've found some other places where a `callable` (array of class name and method) would work but the PHPDoc lists a closure

As a workaround you can create a Closure for pleasing PHP Stan:

~~~php
\Closure::fromCallable([MyClass::class, 'method']);
~~~

I guess a PR fixing the PHPDocs where a `callable` can be accepted in addition to a Closure would be considered.

**Comment by taylorotwell** [comment]
If someone wants to PR a doc block fix go for it.

**Comment by EdgarSedov** [comment]
> I've found some other places where a `callable` (array of class name and method) would work but the PHPDoc lists a closure
> 
> As a workaround you can create a Closure for pleasing PHP Stan:
> 
> ```
> \Closure::fromCallable([MyClass::class, 'method']);
> ```
> 
> I guess a PR fixing the PHPDocs where a `callable` can be accepted in addition to a Closure would be considered.

Hi, can u please mention those places (at least which u remember) so i can make a single fix for all of them?

**Comment by rodrigopedra** [comment]
Hi @EdgarSedov ,

I did a search using my IDE (PHPStorm) yesterday and found around 470+ docblocks with a `\Closure` type hint within the core framework source code (searched for `@param.*Closure` with regex enabled).

Note that I did a simple search, not all those places will be able to be changed to a `callable` typehint, as lots of the methods that use those docblocks are public and have a PHP type hint on their parameter list, so changing those would be a breaking change.

Also some places are used in serialization which I think might not work with an array callable (not sure on this one, but I would skip changing those).

In short, it is a big undertaking reviewing each of those places. I can give a try in a couple of weeks, I won't have the time to do sooner. But please be my guest if you, or anyone else, are available to work on this earlier.
